src/workerinfo.js

```diff
@@ -0,0 +1,69 @@
+let taskcluster = require('taskcluster-client');
+let assert      = require('assert');
+let debug       = require('debug')('workerinfo');
+
+class WorkerInfo {
+  constructor(options) {
+    assert(options);
+    assert(options.Provisioner);
+    this.Provisioner = options.Provisioner;
+
+    // update `expires` values in Azure at this frequency; larger values give less accurate
+    // expires times, but reduce Azure traffic.
+    this.updateFrequency = '6 hours';
+    this.nextUpdateAt = {};
+  }
+
+  /**
+   * Mark a value as seen, only actually updating the Azure table row
+   * occasionally. This takes the approach of updating the row immediately on
+   * the first call, then not updating until `updateFrequency` has elapsed.
+   * Thus the `expires` value on a row may be out-of-date by `updateFrequency`.
+   *
+   * Note that the cache is never purged of outdated entries; this assumes that
+   * the process is restarted on a daily basis, so there is not too much time
+   * for stale cache entries to accumulate.
+   */
+  async valueSeen(key, updateExpires) {
+    let now = new Date();
+    let nextUpdate = this.nextUpdateAt[key];
+    if (!nextUpdate || nextUpdate < now) {
+      await updateExpires();
+      this.nextUpdateAt[key] = taskcluster.fromNow(this.updateFrequency);
```

Comment:
Do this before `await updateExpires();` to avoid doing it multiple times in parallel...

Notably, wrap `await updateExpires();` with a try/catch, where you reset `this.nextUpdateAt[key]` if an exception is thrown... well, maybe, if we can't update it maybe it's better to send the error to sentry and not cause 500 errors on the queue...